File: examples/tmux.conf.d/upinfo.rb
#
#  upinfo.rb  --  Information about uptime and battery
#


module Tmux

  class <<self

    TIME_UNITS = { s: 60, m: 60, h: 24, d: 7, w: nil, }

    def uptime hide: nil, show_zero: nil
      hide = hide.to_sym if hide
      r = []
      t = (File.read "/proc/uptime").to_i
      TIME_UNITS.each { |u,m|
        t, v = m ? (t.divmod m) : [ nil, t]
        r.unshift v, u if !hide && (show_zero || v.nonzero?)
        hide = nil if u == hide
      }
      r.join
    end

    BATTERY_BASE = "/sys/bus/acpi/drivers/battery/PNP0C0A:00/power_supply"

    def battery= bat
      if bat && !bat.empty? then
        @bat = File.join BATTERY_BASE, bat
        @bat_status   = File.join @bat, "status"
        @bat_capacity = File.join @bat, "capacity"
      else
        @bat = @bat_status = @bat_capacity = nil
      end
    end

    def battery
      return unless @bat
      r = ""
      if (File.read @bat_status) =~ /^Charging/i then
        r << "^"
      end
      c = File.read @bat_capacity
      c.rstrip!
      r << c << "%"
    rescue Errno::ENOENT
      "?%"
    end

    def upinfo sep: nil, **kwargs
      a = [ (uptime **kwargs)]
      b = battery
      a.push b if b
      a.join sep||"  "
    end

  end


  if $0 == __FILE__ then
    self.battery = ENV["BAT"]||"BAT"
    puts upinfo
    puts upinfo hide: :s
    puts upinfo hide: "s"
    puts upinfo hide: :m
    puts upinfo hide: :d, show_zero: true
    puts upinfo show_zero: true
    puts upinfo sep: "|", show_zero: true
  end

end

File: examples/tmux.conf.d/upinfo.lua
--
--  upinfo.lua  --  Information about uptime and battery
--

local M = {}


local function divmod(x, y)
  if y then
    local d = x // y  -- since Lua 5.3; alternative: math.floor(x / y)
    return d, x-d*y
  else
    return nil, x
  end
end

local function file_first(p)
  local f = io.open(p)
  if not f then
    return
  end
  local u = f:read()
  f:close()
  return u:match("%S+")
end

local time_units = {
  { u = "s", m = 60 },
  { u = "m", m = 60 },
  { u = "h", m = 24 },
  { u = "d", m =  7 },
  { u = "w"         },
}


M.uptime = function(props)
  local t, v = math.floor(file_first("/proc/uptime"))
  local r = {}
  local hide = props and props.hide
  local show_zero = props and props.show_zero
  for i, e in ipairs(time_units) do
    t, v = divmod(t, e.m)
    if not hide and (show_zero or v ~= 0) then
      table.insert(r, 1, v..e.u)
    end
    if e.u == hide then
      hide = nil
    end
  end
  return table.concat(r)
end

local battery, batterystat, batterycapa

M.set_battery = function(bat)
  if bat then
    battery = "/sys/bus/acpi/drivers/battery/PNP0C0A:00/power_supply/"..bat
    batterystat, batterycapa = battery.."/status", battery.."/capacity"
  else
    battery, batterystat, batterycapa = nil, nil, nil
  end
end

M.battery = function()
  if not battery then
    return
  end
  local r = {}
  if file_first(batterystat) == "Charging" then
    table.insert(r, "^")
  end
  table.insert(r, file_first(batterycapa) or "?")
  table.insert(r, "%")
  return table.concat(r)
end


M.upinfo = function(props)
  local a = {}
  table.insert(a, M.uptime(props))
  local b = M.battery()
  if b then
    table.insert(a, b)
  end
  return table.concat(a, props and props.sep or "  ")
end


--[[
    -- Examples
    ui = require("upinfo")
    ui.set_battery(os.getenv("BAT") or "BAT")
    print(ui.upinfo())
    print(ui.upinfo({hide="s"}))
    print(ui.upinfo({hide="m"}))
    print(ui.upinfo({hide="d", show_zero=true}))
    print(ui.upinfo({show_zero=true}))
    print(ui.upinfo({sep="|", show_zero=true}))
--]]


return M

